File: CarlaDashBoard/bridge_server/bridge/telemetry.py
"""Builds the telemetry JSON envelope published on /stream (design doc §02)."""

import time


def build(session, mode, connected=True):
    ego = session.ego
    if ego is None or not ego.is_alive:
        return {
            "type": "telemetry",
            "t": time.time(),
            "mode": mode.name if mode else "IDLE",
            "connected": connected,
            "ego_alive": False,
        }

    transform = ego.get_transform()
    velocity = ego.get_velocity()
    speed_kmh = 3.6 * (velocity.x ** 2 + velocity.y ** 2 + velocity.z ** 2) ** 0.5
    control = ego.get_control()

    payload = {
        "type": "telemetry",
        "t": time.time(),
        "mode": mode.name if mode else "IDLE",
        "connected": connected,
        "ego_alive": True,
        "speed_kmh": round(speed_kmh, 2),
        "steer": round(control.steer, 4),
        "throttle": round(control.throttle, 4),
        "brake": round(control.brake, 4),
        "x": round(transform.location.x, 2),
        "y": round(transform.location.y, 2),
        "yaw": round(transform.rotation.yaw, 2),
        "town": session.current_town_short(),
    }
    if mode is not None:
        payload.update(mode.status_extra())
    return payload

File: CarlaDashBoard/CarlaDashBoard.Wpf/Models/Telemetry.cs
using System.Text.Json.Serialization;

namespace CarlaDashBoard.Wpf.Models;

/// <summary>
/// Mirrors the "telemetry" text frame published on /stream — design doc §02. The bridge
/// speaks snake_case JSON (Python convention); explicit <see cref="JsonPropertyNameAttribute"/>
/// on every field is the honest mapping instead of a guessed naming policy.
/// </summary>
public sealed class Telemetry
{
    [JsonPropertyName("t")] public double T { get; set; }
    [JsonPropertyName("mode")] public string Mode { get; set; } = "IDLE";
    [JsonPropertyName("connected")] public bool Connected { get; set; }
    [JsonPropertyName("ego_alive")] public bool EgoAlive { get; set; }
    [JsonPropertyName("speed_kmh")] public double SpeedKmh { get; set; }
    [JsonPropertyName("steer")] public double Steer { get; set; }
    [JsonPropertyName("throttle")] public double Throttle { get; set; }
    [JsonPropertyName("brake")] public double Brake { get; set; }
    [JsonPropertyName("x")] public double X { get; set; }
    [JsonPropertyName("y")] public double Y { get; set; }
    [JsonPropertyName("yaw")] public double Yaw { get; set; }
    [JsonPropertyName("town")] public string Town { get; set; } = "";

    // Data Collection mode extras (status_extra() in bridge/modes/data_collection.py)
    [JsonPropertyName("recording")] public bool? Recording { get; set; }
    [JsonPropertyName("collision_count")] public int? CollisionCount { get; set; }
    [JsonPropertyName("lane_invasion_count")] public int? LaneInvasionCount { get; set; }
    [JsonPropertyName("frames_recorded")] public int? FramesRecorded { get; set; }
    [JsonPropertyName("record_elapsed_s")] public double? RecordElapsedS { get; set; }

    public static readonly Telemetry Empty = new() { Mode = "IDLE", Connected = false };
}
